### ISSUE TITLE
Realistic Calculation of Elevation

### ISSUE THREAD

**Issue by y20k** [y20k/trackbook#99]
Trackbook sums up changes in the [altitude](https://developer.android.com/reference/android/location/Location#getAltitude()) information it gets for every location fix. It stores upwards and downwards changes in [positiveElevation](https://github.com/y20k/trackbook/blob/v2.0.10/app/src/main/java/org/y20k/trackbook/core/Track.kt#L45) and in [negativeElevation](https://github.com/y20k/trackbook/blob/v2.0.10/app/src/main/java/org/y20k/trackbook/core/Track.kt#L46). The calculation is done in [calculateElevationDifferences](https://github.com/y20k/trackbook/blob/v2.0.10/app/src/main/java/org/y20k/trackbook/helpers/LocationHelper.kt#L209).

The problem is that altitude values are less reliable than the latitude and longitude values from GPS. Simply summing up the altitude differences between each current location and its predecessor results in unrealistically high elevation values. [calculateElevationDifferences](https://github.com/y20k/trackbook/blob/v2.0.10/app/src/main/java/org/y20k/trackbook/helpers/LocationHelper.kt#L209) drops altitude changes that it determines to be unrealistic. But still the resulting elevation values for a track recording are way too high.

Has anyone here an idea how to get sensible elevation information for a given track?

**Comment by y20k** [comment]
As result of #96 the density of waypoints Trackbook records is increased. The higher density highlights the problems the `calculateElevationDifferences` function.

**Example:** A hilly route in Stuttgart that I am often choosing for a run.

_A. with the previous low density of waypoints_
Elevation (uphill): 158m
Elevation (downhill): -300m
Highest Waypoint: 503m
Lowest Waypoint: 354m

_B. with the new improved density of waypoints_
Elevation (uphill): 461m
Elevation (downhill): -575m
Highest Waypoint: 506m
Lowest Waypoint: 353m

**Comment by voussoir** [comment]
I have two ideas for this. One short one and one long one:

Short idea: Instead of using every single point in the altitude-summing loop, just use 1 point per n meters. Unless the user is climbing ladders up and down, their altitude every 30, 50, or 100m probably represents their track pretty well. I've never paid attention to altitude so I do not have the same personal experience you do about your hilly running path. Pseudocode

```
last_measured_point = null
for point in points
    if last_measured_point == null or track paused and restarted
        last_measured_point = point
    else if distance(point, last_measured_point) > 50m
        add altitudes
        last_measured_point = point
```

---

Long idea: Maybe using a rolling average or line of best fit would help smooth the altitude points before summing. The rolling average could be the average of last n meters.

When I use [GPSTest](https://f-droid.org/en/packages/com.android.gpstest.osmdroid/), I see that there is a vertical accuracy in addition to the horizontal accuracy.

![Screenshot_20210320-165813](https://user-images.githubusercontent.com/7299570/111888867-b38a7b00-899d-11eb-9663-d45bffbeb6b7.png)

According to GPSTest's help screen:

> **E H/V Acc** - Estimated horizontal and vertical accuracy of the location, both at 68% confidence. Vertical accuracy is available on Android 8.0 and higher.

I inspected one of the .json files produced by Trackbook and it seems it does not store this vertical accuracy value. It may be helpful in calculating statistical significance of point-to-point differences if we think this is the right way to go.

But I do think the simple technique of only counting 1 point per n meters would be a good first experiment.

**Comment by Hob111** [comment]
There are two separate issues here, accuracy and precision, simply defined as:- 

**Accuracy**:- the deviation from the true value.
**Precision**:- the repeatability of a measurement.

The accuracy (as defined above) of the altitude on my phone is poor, always being around 40m too high. Therefore this is a _systematic_ error, and is present in all measurements. Hence **any point** on a track will always we out by this amount however it's calculated, and I doubt very much that this will be reflected in the _accuracy_ returned using GPS data. (I will give GPSTest a try to see what I get). I'm sure many other phones will be no better than mine but maybe more expensive devices will have much smaller systematic errors. 

However, when calculating the uphill and downhill elevation of a track we are concerned with relative values and therefore the precision of the measurement rather than the absolute accuracy (_ie_ we look at differences between subsequent readings). The precision of elevation data appears to be poor, (a plot can sometimes look like "a bag of string"), and this is probably what leads to the observation made by @y20k.

> Example: A hilly route in Stuttgart that I am often choosing for a run. _etc....._

Consequently I don't think

> Short idea: Instead of using every single point in the altitude-summing loop, just use 1 point per n meters of horizontal movement. 

will achieve very much, being still subject to the "noise" inherent in the measurements. One of:

> Long idea: Maybe using a rolling average or line of best fit would help smooth the altitude points before summing. The rolling average could be the average of last n meters.

> It may be helpful in calculating statistical significance of point-to-point differences if we think this is the right way to go.

would be more likely to produce an acceptable result. 

To summarise this long post.

- _Either_ calculate a rolling average over, say, 15m
- _Or_ use a full statistical approach to look at the difference between adjacent points.

The first seems the simplest to trial.

**Comment by Hob111** [comment]
> calculate a rolling average over, say, 15m

On further thought, it would be better to calculate a rolling average over the last 10 (say) readings. This would give a more consistent result independent of the velocity of the user, a cyclist could cover 15m in 1s!

**Comment by y20k** [comment]
@voussoir 

> When I use GPSTest, I see that there is a vertical accuracy in addition to the horizontal accuracy.

Ah interesting. I looked it up. Since API 26 `Location` has a vertical accuracy attribute the can be read via [getVerticalAccuracyMeters()](https://developer.android.com/reference/kotlin/android/location/Location#getverticalaccuracymeters). If we want to use this, the required OS version for Trackbook would be Android 8 (Oreo) - it currently is Android 7.1 (Nougat).

**Comment by Hob111** [comment]
I installed GPSTest and got the expected difference between my actual altitude and that reported by the phone. However, a little more research reveals that this isn't an error produced by my phone but rather because GPS doesn't directly give the altitude above mean sea level (MSL) (used on UK maps and I imagine many others), but instead all measurements are based on an _oblate spheroid_ with its centre at the centre of the Earth, the [World Geodetic System (WGS84)](https://gisgeography.com/wgs84-world-geodetic-system/) and [also on Wikipedia](https://en.wikipedia.org/wiki/World_Geodetic_System).

So reported altitudes will always differ from those based on MSL but the differences needed to calculate the uphill and downhill elevation are unaffected.

**Comment by Hob111** [comment]
@y20k When calculating uphill and downhill elevation I think what you are doing is integrating all reported increases in altitude and separately integrating all reported decreases in altitude. This means you are also integrating all of the positive and negative errors, so it's not surprising that the more points you take the larger the uphill and downhill elevations become.

If the vertical accuracy is used to filter out bad data I don't think this addresses the basic problem, it will remove points with poor vertical accuracy but you'll still be adding together all the negative errors and positive errors on the remaining points.

However, a running average calculated over a pre-set number of points will effectively smooth the altitude curve and remove much of the noise, leading to a much more realistic estimate of the uphill and downhill elevation.

**Comment by voussoir** [comment]
> If we want to use this, the required OS version for Trackbook would be Android 8 (Oreo) - it currently is Android 7.1 (Nougat).

Rather than bumping the minimum supported version, we could have a separate code path for when vertical accuracy is not present. On the previous issue for example, we had to handle the case that accuracy was 0, so we could do the same with vertical accuracy, just let it be zero on Android < 8.

> However, a running average calculated over a pre-set number of points will effectively smooth the altitude curve and remove much of the noise, leading to a much more realistic estimate of the uphill and downhill elevation.

I agree that running a smoother / rolling average would be a good approach before we get into a full-on statistical approach.

**Comment by y20k** [comment]
Just for me to understand this ... (I am not familiar with the term rolling average). 

1. for each new waypoint x
2. we take the previous 9 waypoints
3. calculate the average altitude of those 10 waypoints
4. ... by just adding them together and dividing them by 10?
5. save the average altitude to waypoint x
6. elevation is calculated as before - by comparing altitude of waypoint x with its predecessor

**Comment by voussoir** [comment]
@y20k Yes that's right. I would also consider doing a rolling average based on distance rather than number of points. That is, for each waypoint x, take all previous waypoints where `distance(x, y) < 30` or 50 or 100 or something. That way the point density doesn't play as much of a role.

I'm sure there are better smoothing techniques than a rolling average to get a line of best fit, but I'd have to research that as I don't remember them at the moment.

**Comment by Hob111** [comment]
> I agree that running a smoother / rolling average would be a good approach before we get into a full-on statistical approach.

It is the only way to tackle this problem. As I've tried to explain previously, a statistical approach doesn't tackle the underlying problem that the positive and negative errors are being summed separately. A statistical approach will only remove any ill defined altitudes, it won't solve anything else.

**Comment by Hob111** [comment]
> Just for me to understand this ... (I am not familiar with the term rolling average).

The term I prefer is moving average, I think it's more descriptive. For example, with a period of 10, the first point you'd use would be the average of waypoints 1 to 10, the second the average of waypoints 2 to 11, then 3 to 12, 4 to 13 etc.

**Comment by Hob111** [comment]
> @y20k Yes that's right. I would also consider doing a rolling average based on distance rather than number of points. That is, for each waypoint x, take all previous waypoints where distance(x, y) < 30 or 50 or 100 or something. That way the point density doesn't play as much of a role.

I must disagree with this, using distance imposes an artificial structure on the data and isn't mathematically rigorous. As I've said previously, if 15m was chosen as the distance a cyclist could cover that in 1s, so you'd be taking a moving average every 1s, ie no average at all, whereas for a walker you'd only be averaging every10th waypoint, so a lot of the data would be ignored!

The only rigorous approach is to average the altitude for every waypoint that has been accepted by the statistical check on distance moved. This is independent of the velocity of the user and uses all of the data, essential if the result is to be meaningful.

**Comment by Hob111** [comment]
> I'm sure there are better smoothing techniques than a rolling average to get a line of best fit, but I'd have to research that as I don't remember them at the moment.

A moving average is very powerful way of smoothing date that shows "noise". The only enhancement that come to mind immediately is to use a weighted moving average, where each point is given a weight dependent on some measure of its reliability. In this case the error on each altitude reading could be used in some way. I think this is an unnecessary complication for this problem.

**Comment by y20k** [comment]
@voussoir @Hob111 just a side note: I love the discussions here and on #96 👍 ❤️

**Comment by Hob111** [comment]
Me too. It's good to get my brain going on a few problems.

**Comment by y20k** [comment]
I wrote a small function to calculate the moving/rolling average. I need to wire it up in the app and test it. The function currently looks like this:

```
    private fun calculateAltitude(location: Location, track: Track): Double {
        val trackSize: Int = track.wayPoints.size
        // skip calculation if less than two waypoints available
        if (trackSize < 2) return location.altitude
        // get number of locations to be used in calculating the moving average
        val numberOfLocationsUsedForSmoothing: Int = if (trackSize < Keys.DEFAULT_NUMER_OF_LOCATIONS_USED_FOR_SMOOTHING) {
            trackSize
        } else {
            Keys.DEFAULT_NUMER_OF_LOCATIONS_USED_FOR_SMOOTHING
        }
        // add altitude values in range and calculate average
        val mostRecentWaypointIndex: Int = trackSize - 1
        var altitudeSum: Double = 0.0
        for (i in mostRecentWaypointIndex..(mostRecentWaypointIndex - numberOfLocationsUsedForSmoothing)) {
            altitudeSum = altitudeSum + track.wayPoints[i].altitude
        }
        return altitudeSum / numberOfLocationsUsedForSmoothing
    }
```
When/if it get it working I will share an APK.

**Comment by voussoir** [comment]
@y20k Awesome. It looks like this function is run for every point that is being added to the track, is that right? To improve performance, it might be beneficial to keep a list / deque of the previous n points, and maintain an ongoing sum so you only have to add the newest and remove the oldest on every iteration. This way, you do not have to perform the entire `for` loop for each point. [Wikipedia describes](<https://en.wikipedia.org/wiki/Moving_average#Simple_moving_average_(boxcar_filter)>):

> When calculating the next mean SMA_knext with the same sampling width k [...] A new value p_n+1 comes into the sum and the oldest value p_n-k+1 drops out. This simplifies the 
calculations by reusing the previous mean SMA_kprev.

> SMA_knext = SMA_kprev + (1/k)(p_n+1 - p_n-k+1)

> This means that the moving average filter can be computed quite cheaply on real time data with a FIFO / circular buffer and only 3 arithmetic steps.

However, I know that introducing global / persistent state increases the chance of bugs elsewhere, since the deque + sum would need to be cleared when the track is ended, paused, etc. Looking forward to this progress.

**Comment by Hob111** [comment]
@y20k Great! It's not obvious to me how many points you're using in the moving average, I assume it's `Keys.DEFAULT_NUMER_OF_LOCATIONS_USED_FOR_SMOOTHING` and it will be set elsewhere, or have I missed something or got it wrong? 

I mention this because if it's set too low it won't remove all of the "noise" whereas if set too high you could "cut off" the tops of hills and bottoms of valleys. 

Unless @voussoir knows of a simple way to determine the optimum number of points I suggest starting with what seems a reasonable value (10?) and then comparing experiment with theory. There is a UK web site where I can upload a track and recalculate the ascent and descent, which I assume is based on map contours. Are there similar sites in your countries? 

Values within about 10% of the calculated figures would seem to me to be very acceptable. Anyone got any other views or better ideas?

**Comment by voussoir** [comment]
I've been doing some experimenting. I generated some **[fictional data](https://gist.github.com/voussoir/7943283c4417bec40e9320d5c0f84bd6)** simulating a 2km walk over some unusual terrain, and tried different smoothing techniques. Please notice the y axis. These two samples have altitude spans of 50/80 meters, so these grades might be unnaturally steep, but again I don't focus on elevation when I walk so I don't know.

The gray points are raw, simulated data and the black points are different smoothing techniques. Please click on the images to open them in a new tab and zoom.

![ca](https://user-images.githubusercontent.com/7299570/112544314-156b2c00-8d74-11eb-9bcb-f19fbb7e8335.png)
![cb](https://user-images.githubusercontent.com/7299570/112544315-156b2c00-8d74-11eb-9301-73a6dd434731.png)
![cc](https://user-images.githubusercontent.com/7299570/112544316-1603c280-8d74-11eb-887e-5f2ab37826de.png)
![da](https://user-images.githubusercontent.com/7299570/112544318-1603c280-8d74-11eb-8101-09f2d41c8d64.png)
![db](https://user-images.githubusercontent.com/7299570/112544320-1603c280-8d74-11eb-9b70-0564e96a700f.png)
![dc](https://user-images.githubusercontent.com/7299570/112544321-169c5900-8d74-11eb-8f22-5cd845c951a7.png)

In general, averaging by 10 points is good and produces smooth lines. But, if the user is traveling quickly, or has poor GPS such that points are far apart, n-based smoothing causes the points to lag. Raising n to 25 points causes extreme lag, so that's no good.

Now, I was only summing from left to right to get the average, but it's also possible to sum on both sides of the target point, half on each side, which would probably eliminate the lag. But, that would have to be done retroactively whereas this can be done in real time.

But I think distance-based smoothing works very well! @Hob111,

> > I would also consider doing a rolling average based on distance rather than number of points. That is, for each waypoint x, take all previous waypoints where distance(x, y) < 30 or 50 or 100 or something. That way the point density doesn't play as much of a role.

> if 15m was chosen as the distance a cyclist could cover that in 1s, so you'd be taking a moving average every 1s, ie no average at all, whereas for a walker you'd only be averaging every10th waypoint, so a lot of the data would be ignored!

This is absolutely true. But that's why I averaged over 50-100m instead of 15! Unless the user is going over 100km/h, we're still averaging at least a couple of data points, but without the data lag caused by using a fixed number of points. For a walker, the data is not ignored, it is summed into the average.

**Conclusion:** In good conditions, distance-based averaging is just as good as n-based averaging. With 5m resolution and a 100m window, you're averaging 20 points. In bad conditions, distance-based averaging is not as smooth as n-based averaging. With 15m resolution, you're averaging 6 points instead of 10. With 35m resolution, you're only averaging 3 points. But, the results are still reasonably smooth and does not suffer from the lag effect.

But, this is simulated data so it's possible my simulation is stupid.

**Comment by Hob111** [comment]
@voussoir Thanks for looking at this, I've been frustrated by a lack of suitable software!

I'm puzzled why there are fewer smoothed points than data points. The whole purpose of a moving average is that you produce as many smoothed points as the original data points (minus a few at the beginning which have to be dealt with in some way). What have I missed? Are you smoothing a subset of the random data? In which case it would be helpful to have the subset plotted not the whole dataset so we can see exactly what is being smoothed.

> > if 15m was chosen as the distance a cyclist could cover that in 1s, so you'd be taking a moving average every 1s, ie no average at all, whereas for a walker you'd only be averaging every 10th waypoint, so a lot of the data would be ignored!

> This is absolutely true. But that's why I averaged over 50-100m instead of 15! Unless the user is going over 100km/h, we're still averaging at least a couple of data points, but without the data lag caused by using a fixed number of points. For a walker, the data is not ignored, it is summed into the average.

Averaging over 50-100m will lose a small hill! If you use all of the data "n-based" smoothing is the only way to approach this without imposing some selection process on the data. The object isn't to produce a totally smooth line but to remove "noise" without destroying detail, imposing a pre-selection could well destroy detail. Each smoothed point **must** be created from the same number of original data points. The trick is to choose "n" carefully, I suspect we'll find a value of less than 10 is suitable.

As I understand it the purpose of smoothing the data is two-fold:

1. to produce realistic values for the uphill and downhill elevation.
2. to produce a smoother and more realistic altitude profile without too many random fluctuations. I'm assuming the smoothed altitudes, not the unsmoothed raw data, will be written to the gpx file (or the file Trackbook uses to hold the track). (This has implications for #85).

Inevitably smoothing will produce some lag, and the larger _n_ the greater the lag. It's important to choose  a value for _n_ as small as possible to give the required result, this also ensures the data aren't over-smoothed and detail is lost.
 
> Now, I was only summing from left to right to get the average, but it's also possible to sum on both sides of the target point, half on each side, which would probably eliminate the lag. But, that would have to be done retroactively whereas this can be done in real time.

Not necessarily. 

1. Choose an odd value for _n_.
2. Save the smoothed value for points 1 to _n_ with the location values for point _n_-(_n_-1)/2
3. Save the smoothed value for points 2 to _n_+1 with the location values for point (_n_+1)-(_n_-1)/2

In general, for points _i_-_n_+1 to _i,_ save the smoothed value with location values for point _i_-(_n_-1)/2

I assume such manipulations are possible in Java, although for relatively small values of _n_ this should be unnecessary.

**Comment by Hob111** [comment]
I though an illustration of what we're aiming for might be useful, so a profile from a walk I did last year.

**Original Trackbook data**  Altitudes are based on GPS values (see my post some way above)
![Original Trackbook data](https://user-images.githubusercontent.com/38910411/112677658-f0ed7d80-8e61-11eb-885f-8b9ca46b1479.png)
**Calculated profile for the same track.** Altitudes based on MSL
![Calculated data for same track](https://user-images.githubusercontent.com/38910411/112677714-fe0a6c80-8e61-11eb-96d8-b68b005842fb.png)

**Comment by voussoir** [comment]
> I'm puzzled why there are fewer smoothed points than data points. Are you smoothing a subset of the random data?

Yes, I generated random data for every meter, and then for each set of graphs took subsets of 1 point per 5 meters, 15m, 35m to simulate varying GPS conditions.

> Averaging over 50-100m will lose a small hill!

If the user's Trackbook records a point once every 10 meters, then using a 100m distance-based window or a 10 point n-based window will produce identical results. I don't think it's right to say that distance-based windows inherently remove more detail. My algorithm doesn't eliminate points, it includes them in the average. The difference is that using a fixed n-based window will cause lag under poor conditions, whereas the distance-based window will use many points in good conditions and fewer points in bad conditions, but it's always better than nothing.

I can see that each technique has its strength and weakness. I'll be happy with whatever @y20k chooses to implement, but I feel that we just have a difference in preference.

**Comment by Hob111** [comment]
Hi @voussoir 

We have different preferences indeed, and I'm trying to make sure I understand exactly what you're suggesting. I believe your method has the potential to lose detail but I want to make sure I'm understanding correctly.

> If the user's Trackbook records a point once every 10 meters, then using a 100m distance-based window or a 10 point n-based window will produce identical results.

I assume it follows that if Trackbook records a point every 5 metres the 100m distance based window will include 20 points averaged. Whereas the 10 point window will span 50 metres. So the better the data the more the distance window puts together in an average.  However, if the data are poor, say 1 point every 20 metres, there will be five points averaged in the 100m window but the 10 point window will span 200m. so each distance window may carry a different number of points, weighting the points unequally and seeming to give higher weight to less good data. With the n-based data the best data will give maximum detail and the poor data the least detail, as it ought.

Raw data is unlikely to be of a consistent quality and it is desirable that in good signal areas the extra information contained in the data should be used.

> The difference is that using a fixed n-based window will cause lag under poor conditions,

This is trivial unless n is large or the data are very poor, and if it is an issue it's possible to avoid as I explained above.

On balance I go for minimising loss of detail.

**Comment by Hob111** [comment]
@y20k 

> I wrote a small function to calculate the moving/rolling average. I need to wire it up in the app and test it. The function currently looks like this:

If you manage to implement this would it be a lot more work to provide a setting so that the number of points in the moving average can be changed? We need to try various values to find one which best removes the "noise" without smoothing/removing genuine detail. I'm only suggesting this for a trial version not the final released app.

> When/if it get it working I will share an APK.

A setting as described above would avoid having to produce numerous APKs.

**Comment by y20k** [comment]
@y20k 
Just a short sign of life. I have quite busy the last couple of days. But will definitely implement some kind of smoothing to the altitude data. I am annoyed with the results of the current production version of Trackbook. It just might take a couple of weeks until I find the time.

**Comment by y20k** [comment]
@voussoir @Hob111 
Today I finally had a couple hours to work on this. I implemented settings for Altitude Smoothing and for #96. And I implemented the code for altitude smoothing. See [commit](https://github.com/y20k/trackbook/commit/3fa589e21c45edca51ff5ca9bed3b46f248122d0).

<a href="https://user-images.githubusercontent.com/9103935/116818851-c16f1680-ab6d-11eb-89cd-af48062f0edf.png"><img src="https://user-images.githubusercontent.com/9103935/116818851-c16f1680-ab6d-11eb-89cd-af48062f0edf.png" width="240"></a>

I haven't tested this yet. I will do so next week. After I made sure, that I did not make any catastrophic errors, I will post an APK for you to test this, too.

**Comment by voussoir** [comment]
@y20k I was very happy to see this in my inbox this morning. Thanks so much for that commit!

From reading the code, I expect it will work. My understanding is that the corrected altitude does not actually overwrite the waypoint's `location.altitude`, it is only used to calculate the sum positive and negative elevation totals. Looks good to me.

I see a few opportunities to improve code performance. Rolling averages can be calculated by maintaining a queue and a sum. When you add a node, you add its value to the sum; when you pop a node, you remove its value from the sum. I did this in my [experiment](https://gist.github.com/voussoir/7943283c4417bec40e9320d5c0f84bd6#file-altitude-py-L30). Of course, this would require resetting the queue and sum when starting a new track, unpausing the track, etc.

The Trackbook code is doing a for-loop of n points for every single point that is added. And:

```Kotlin
val locationAltitudeCorrected: Double = calculateCorrectedAltitude(location, track, altitudeSmoothingValue)
val previousLocationAltitudeCorrected: Double = calculateCorrectedAltitude(previousLocation, track, altitudeSmoothingValue)
```

This n-loop is occuring twice per point, where 14 out of 15 points in these sums will be identical. I guess this is small but might be a place to revisit later.

One more thing I noticed:

```Kotlin
val elevationDifferences: Pair<Double, Double> = LocationHelper.calculateElevationDifferences(previousLocation, location, track, altitudeSmoothingValue)
// check if any differences were calculated
if (elevationDifferences != Pair(track.positiveElevation, track.negativeElevation)) {
    // update elevation values (do not update if resumed -> we do not want to add values calculated during a recording pause)
    if (!resumed) {
        track.positiveElevation = elevationDifferences.first
        track.negativeElevation = elevationDifferences.second
    }
}
```

Can `if (!resumed)` be moved higher up, so that it doesn't calculate the difference in the first place? It seems the value is thrown away.

Looking forward to the next release!

**Comment by y20k** [comment]
@voussoir 

I tested the solution earlier. It still has bug. Elevation values were both zero. 

> Rolling averages can be calculated by maintaining a queue and a sum. 

I agree that would be a lot more efficient. I guess it makes sense to store the queue in TrackerService. I will look into it.

**Comment by y20k** [comment]
Just a short update. I now added a class that holds the 
- last altitude values in a list
- a sum of the last altitude values
- an average

https://github.com/y20k/trackbook/blob/bb00e18312cfc07b3515139ac76c8bba6a3de7f6/app/src/main/java/org/y20k/trackbook/TrackerService.kt#L563-L575

The approach does not work yet. But conceptually I like it already better.

**Comment by voussoir** [comment]
Sounds great 👍 I'm writing from mobile now but I'd like to mention:

1. Instead of storing the average as a member of the class, I'd suggest adding a method to recalculate average when needed.
2. `sum += element` shouldn't be in else since it should always happen when adding an element.

Thanks @y20k!

**Comment by Hob111** [comment]
Hi you two, 

I'm currently on a walking holiday in N England. I'll contribute to this discussion when I return home in a week's time.

**Comment by y20k** [comment]
@voussoir 
Thank you for the quick look. Indeed the sum was not calculated correctly 👍 👍

**Comment by y20k** [comment]
Hi there,

I conducted two tests using the new smoothing function. If you remember the whole issues started with me noticing the elevation values changing with the higher density of waypoints that are a result of Issue #96. Here the relevant part of my comment back then.

> A. with the previous low density of waypoints
> Elevation (uphill): 158m
> Elevation (downhill): -300m
> Highest Waypoint: 503m
> Lowest Waypoint: 354m
> 
> B. with the new improved density of waypoints
> Elevation (uphill): 461m
> Elevation (downhill): -575m
> Highest Waypoint: 506m
> Lowest Waypoint: 353m

And here are now the results for _two_ runs so far. I used the high accuracy recording setting (= [Factor 2](https://github.com/y20k/trackbook/issues/96#issuecomment-830840926)) from #96. And I set the altitude smoothing slider to **10**.

Elevation (uphill): 120m | -115m
Elevation (downhill): -242m | -247m
Highest Waypoint: 504m | 508m 
Lowest Waypoint: 348m | 354m

I will give it another shot next week and report back - this time with the slider set to 15. I reserve a judgment for after that test. But I must say that the results so far look very promising.

Anyway. Since the tests I conducted went well. I did not experience any problems... here is a signed APK built on the code up to the commit [5fe4b8d](https://github.com/y20k/trackbook/commit/5fe4b8d13e2f5e8d1c1449a0385801af4533c49e)

- [trackbook-app-testing-2.1.0-003.apk](https://drive.google.com/file/d/1mGgW2SRfCS0lyHKyAo-3gmzDbJQ8M2O6/view?usp=sharing)

**Comment by Hob111** [comment]
> @voussoir From reading the code, I expect it will work. My understanding is that the corrected altitude does not actually overwrite the waypoint's location.altitude, it is only used to calculate the sum positive and negative elevation totals. Looks good to me.

This is a simple way of avoiding any "lag" should the number of points averaged get too large but won't help if #85 is implemented, when the altitude graph will look like a porcupine's/hedgehog's back 😄 . 

> @y20k And here are now the results for two runs so far. I used the high accuracy recording setting (= Factor 2) from #96. And I set the altitude smoothing slider to 10.

We need to be careful about setting the number of points smoothed too high. The object is to remove high frequency noise (the random variation or errors on each altitude) from the lower frequency signal (the actual variations in altitude).  The more points that are smoothed the more the signal is affected, 

> Anyway. Since the tests I conducted went well. I did not experience any problems... here is a signed APK built on the code up to the commit 5fe4b8d

Excellent. I'll try comparing my results with calculated values to see how close they are. I'll post anything I get here. Thanks for all your hard work @y20k

**Comment by Hob111** [comment]
I agree that most users wouldn't understand the purpose of the slider. I think my concern originates in my scientific/mathematical background, _"always keep the problem you are trying to solve at the front of your mind"_. The noise is high frequency so, in my opinion, if we cant get a satisfactory result by smoothing ten or fewer points then we've not found a suitable solution and are simply fudging a result, suggesting we maybe need to look at a weighted moving average of some kind, which could start getting horribly complicated  :grimacing:.

The other point we should think about is whether the multiple of sigma we use (#96) has an affect on the optimum number of points we need to smooth. This is probably best checked by experiment.

**Comment by Hob111** [comment]
I've conducted my first test of trackbook-app-testing-2.1.0-003.apk ("See Emily Play" - love the name, did I miss "Arnold Lane"?).

The results are promising.

Walk: 14.1km. Circular, so ascent and decent the same. 
Conditions: Very cloudy, continuous rain for the second half of the walk.
Terrain: Mostly open countryside, a few wooded areas.
Smoothing factor: 6. Recording Accuracy: High (2-sigma).
Elevation reported by Trackbook: 360m
Elevation from Trackbook gpx output file: 690m
Calculated elevation: 236m

So the error in the elevation has been reduced from circa 190% to 50% (Note: I don't know what the error might be on the calculated value, I assume quite small.)

I'll do another test later this week using a smoothing factor of 7.

**Comment by y20k** [comment]
I recorded another two runs. This time setting the smoothing factor to **15**. And I put all results so far into the table below for better comparison.

|             | Run 1 | Run 2 | Run 3 | Run 4 | Run 5 | Run 6 | 
| ----------- | ----- | ----- | ----- | ----- | ----- | ----- | 
| Elev. up    | 158m  | 461m  | 120m  | 115m  | 79m   | 109m  | 
| Elev. down  | -300m | -575m | -242m | -247m | -184m | -238m |
| Highest Pt. | 503m  | 506m  | 504m  | 508m  | 517m  | 505m  |
| Lowest Pt   | 354m  | 353m  | 348m  | 354m  | 372m  | 357m  |

Run 1: no improvements (= current Trackbook v2.0.10)
Run 2: increased waypoint density (#96) (Factor 1), no altitude smoothing

Run 3: increased waypoint density (Factor 2), altitude smoothing set to **10**
Run 4: increased waypoint density (Factor 2), altitude smoothing set to **10**

Run 5: increased waypoint density (Factor 2), altitude smoothing set to **15**
Run 6: increased waypoint density (Factor 2), altitude smoothing set to **15**

**Comment by Hob111** [comment]
Hi @y20k. I assume these are the same route. If so, 3 and 4 are about what I'd expect, just a bit of random variation. However, 5 and 6 show surprising differences! 5 is about what one might expect but 6 isn't a lot different to 3 and 4! Very odd. 

 For 3,4,5, and 6 you say "increased waypoint density (Factor 2)". I interpret that to mean you've used the app with the statistical evaluation and 2-sigma.

I didn't report the highest or lowest points because I didn't know if these were smoothed or unsmoothed values. Your values seem fairly consistent so I assume unsmoothed.

**Comment by y20k** [comment]
Hi @Hob111 

>  For 3,4,5, and 6 you say "increased waypoint density (Factor 2)". I interpret that to mean you've used the app with the statistical evaluation and 2-sigma.

Yes. That is what I meant :-)

> I didn't report the highest or lowest points because I didn't know if these were smoothed or unsmoothed values. Your values seem fairly consistent so I assume unsmoothed.

That is correct, too. The highest and lowest points are not smoothed. Not sure why I reported them at all to be honest. I guess similar values here may hint to similar weather conditions, if altitude values are calculated using a barometer. But I think we can ignore them.

PS. I added a Run 7 to the table above.

**Comment by mike-of-earth** [comment]
Just popping in to say THANK YOU!!!! This is the conversation that other apps I use should be having. Instead, they're just out for building a social platform with recording features.

Anyway, I've downloaded the test apk you posted above and will test it out on my mountain bike here in Colorado. Elevation gain and total time are my to most important metrics, and I've struggled getting accurate elevation numbers from another GPS recording app.

I will compare the trackbook elevation results with other recordings I have on some trails that I have ridden a ton and know the elevation.
Thanks again for your hard work... and for a great app!

**Comment by y20k** [comment]
Sounds like a good plan. I admit I felt I had no way to confidently decide on a default smooting value. Looking forward to your findings!!!

**Comment by mike-of-earth** [comment]
Here with my completely unscientific results. I've ridden a bunch of my typical rides here and have been comparing my results against 'known' elevation gains. I say known because I'm not sure how they were established originally. I've not used google earth or another topo app to compare my results, but I'm pretty confident in the numbers I'm getting in trackbook.

A setting of 13 is perfect. 10 counted too many points on trails with lots of small changes in elevation. 13 smoothed those out without dropping relevant events.

**Comment by y20k** [comment]
Hi @mike-of-earth 
I will try the setting 13 during my runs, too. I find nothing odd with the results I will remove the "Smoothing" slider from Settings, package up the current state of the app and publish a new version. Thanks for your tests!!!

**Comment by y20k** [comment]
I finally found the time to publish a new version containing the enhancements we came up with in #99 and #96. 🥳  ([v2.1.0](https://github.com/y20k/trackbook/releases/tag/v2.1.0)) 

A huge THANK YOU for your help here
